docs/enhancements/dual-stack-vips.md

```diff
@@ -0,0 +1,114 @@
+
+## Proposal
+
+### User Stories
+
+* As an operator of the IPv6-only Hub cluster I want to deploy dual-stack Spoke cluster.
+
+* As an user of a dual stack OCP cluster I want to run IPv6-only applications that need access to the API and/or Ingress VIPs.
+
+### API Extensions
+
+New fields will have to be added to the Cluster object to persist the specified VIPs.
+
+### Implementation Details
+
+Our change can be simply described as "making a singular field plural". For this, we will follow the [Kubernetes API change guide](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#making-a-singular-field-plural).
+
+#### Phase 1
+
+`cluster-create-params`, `v2-cluster-update-params` and `cluster` will get new fields
+
+* `api_vips` of type `array`
+* `ingress_vips` of type `array`
+
+Respective AgentClusterInstall CR will also get those new fields.
+
+The database schema will be updated by creating new tables `api_vips` and `ingress_vips` with the following columns
+
+* `cluster_id`
+* `vip`
+
+#### Phase 2
+
+A set of functions for cluster creation will provide a compatiblity between already existing fields (`api_vip` and `ingress_vip`) and new ones (`api_vips` and `ingress_vips`). The behaviour will be as follows
+
+* If only `api_vip` is provided, the `api_vips` will be automatically populated by a simple conversion from `string` to `[]string`. This guarantees no change in behaviour for users not aware of introduction of the new field.
+
+* If both `api_vip` and `api_vips` are provided, the value of `api_vip[0]` must match the value of `api_vip`.
+
+* If only `api_vips` are provided, the error will be raised. This is the case when plural field is specified and the singular is not. The rationale is that in an update it's impossible to tell the difference between an old client clearing the singular field via patch and a new client setting the plural field. The update semantics must not differ from create, therefore behaviour as described.
+
+For cluster update the following logic will apply
+
+* If `api_vip` is cleared and `api_vips` are not provided, we will clear both fields. This is a scenario when an old client (not aware of plural field) is used.
+
+* If `api_vips` is cleared and `api_vip` is not provided, the change will not apply. This is a scenario when an old client is used and it can't send fields it doesn't know about.
+
+* If `api_vip` is changed (but not cleared), `api_vips` will be forcefully set to the value of `api_vip` as a one-element list. This is a scenario when an old client is used to change the field it knows about.
+
+All the points described above will behave in an equivalent way for `ingress_vip` and `ingress_vips`.
+
+#### Phase 3
+
+After providing a required period of backwards-compatibility the `api_vip` and `ingress_vip` fields will be deprecated and removed. The value in the database will be nulled. Removal of the column in the database will follow.
```

Comment:
It will impact not only UI & test-infra but also external tools such as: aicli

Reply:
That's correct, I will write it explicitly. Also Crucible and siteconfig generator will be affected (that's all interfaces that I know about)